logicalplan/regex_matchers.go

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"strings"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/thanos-community/promql-engine/parser"
+)
+
+const hasRegexChars = ".+|[]{}^$*?+()\\"
+
+// RegexMatchersFunctions function will optimize the query by avoiding unnecessary regex match.
+// For example: up{app=~"something"} will be optimized to up{app="something"}.
+type RegexMatchersFunctions struct {
+}
+
+func (RegexMatchersFunctions) Optimize(expr parser.Expr, _ *Opts) parser.Expr {
+	traverse(&expr, func(node *parser.Expr) {
+		switch e := (*node).(type) {
+		case *parser.VectorSelector:
+			for matcherIndex, lm := range e.LabelMatchers {
+				if lm.Type == labels.MatchRegexp {
+					if !strings.ContainsAny(lm.Value, hasRegexChars) {
```

Comment:
This is different to https://github.com/prometheus/prometheus/pull/12434/files#diff-c6bfdf5ae2a5edbc3df844f48f297e4d2a4f4106662f1a04c8931fbb38f1fdaeR83, though.

Reply:
the way in the prometheus pr was the only way i could come up with to make sure that its a literal without parsing though, is it problematic?